src/actions/firebase/firebase.ts

```diff
@@ -0,0 +1,167 @@
+      return new Promise(async (resolve, reject) => {
+        let data: any = params.data
+        if (!params.title) {
+          reject("Needs a valid title.")
+        }
+        if (!params.deviceIds) {
+          resolve()
+        }
+        let notification: any = {title: params.title}
+        let notificationData: any = data
+        notificationData.imageUrl = imageName
+        if (params.timeToLive) {
+          this.notificationOptions.timeToLive = parseFloat(params.timeToLive)
+        }
+        if (params.priority) {
+          this.notificationOptions.priority = params.priority
+        }
+        try {
+          const userObj = JSON.parse(params.deviceIds ?? "[]") as any[]
+          let deviceIdObject = params.deviceIds as unknown as any[]
+          if (deviceIdObject.length == 0) {
+            resolve()
+          }
+          await userObj.forEach(async (userDevices) => {
+            let devices = userDevices as any[]
+            await devices.forEach(async (device, index) => {
+              if (device.device_id && device.user_id) {
+                let deviceId = device.device_id.toString()
+                notificationData.userId = device.user_id.toString()
+                let payload = {
+                  notification: notification,
+                  data: notificationData
+                };
+                try {
+                  await this.sendMessageToDevice(deviceId, payload, this.notificationOptions)
+                }
+                catch (error) {
+                  reject(error)
+                }
+              }
+              if (index === devices.length -1) {
+                resolve()
+              }
+            })
+          });
+        } catch (error) {
+          reject(error)
+        }
+      })
+      .catch((error) => {
+        throw error
+      })
+    }
+
+    async sendMessageToDevice(deviceId: string, payload: admin.messaging.MessagingPayload, options?: admin.messaging.MessagingOptions): Promise<any> {
+      return new Promise((resolve, reject) => {
+        FirebaseAction.setFirebaseClient()
+        admin.messaging().sendToDevice(deviceId, payload, options)
+          .then( (_: any) => {
+            resolve()
+          })
+          .catch( (error: any) => {
+            reject(error.message)
+          });
+      })
+      .catch((error) => {
+        throw error
+      });
+    }
+
+    async uploadImage(request: Hub.ActionRequest): Promise<any> {
+      let data = request.formParams.data as any
+
+      try {
+        if (request.attachment?.dataBuffer) {
+          return new Promise((resolve, reject) => {
+            FirebaseAction.setFirebaseClient()
+            let dateString = new Date().toISOString()
+            let alertId = data.alertId ?? ""
+            const destFileName = dateString + alertId;
+            const fileCloud = admin.storage().bucket(this.storageBucket).file(destFileName);
+              fileCloud.save(request.attachment?.dataBuffer!, {
```

Comment:
couple of questions 
- what is the auth mechanism for accessing this file
- for how long will this file remain in firebase server
- can there be data sovereignty issue

Reply:
Will update the security document and update it here.

Reply:
https://docs.google.com/document/d/1hTpLx7ScTV4yuW5A88iFVkjr4eGe6EUaqZ86oDC6JmU/edit#

Reply:
Have made a doc
Will send it to the reviewers